lighthouse-core/lib/i18n/i18n.js

```diff
@@ -369,143 +354,140 @@ function getRendererFormattedStrings(locale) {
 
   /**
-   * Convert a message string & replacement values into an
-   * indexed id value in the form '{messageid} | # {index}'.
-   *
-   * @param {string} icuMessage
+   * Convert a message string and replacement values into an `LH.IcuMessage`.
+   * @param {string} message
    * @param {Record<string, string | number>} [values]
-   * */
-  const getMessageInstanceIdFn = (icuMessage, values) => {
-    const keyname = Object.keys(mergedStrings).find(key => mergedStrings[key] === icuMessage);
-    if (!keyname) throw new Error(`Could not locate: ${icuMessage}`);
+   * @return {LH.IcuMessage}
+   */
+  const getIcuMessageFn = (message, values) => {
+    const keyname = Object.keys(mergedStrings).find(key => mergedStrings[key] === message);
+    if (!keyname) throw new Error(`Could not locate: ${message}`);
 
     const filenameToLookup = keyname in fileStrings ? filename : __filename;
     const unixStyleFilename = path.relative(LH_ROOT, filenameToLookup).replace(/\\/g, '/');
-    const icuMessageId = `${unixStyleFilename} | ${keyname}`;
-    const icuMessageInstances = _icuMessageInstanceMap.get(icuMessageId) || [];
+    const i18nId = `${unixStyleFilename} | ${keyname}`;
 
-    let indexOfInstance = icuMessageInstances.findIndex(inst => isDeepEqual(inst.values, values));
-    if (indexOfInstance === -1) {
-      icuMessageInstances.push({icuMessageId, icuMessage, values});
-      indexOfInstance = icuMessageInstances.length - 1;
-    }
-
-    _icuMessageInstanceMap.set(icuMessageId, icuMessageInstances);
-
-    return `${icuMessageId} # ${indexOfInstance}`;
+    return {
+      i18nId,
+      values,
+      formattedDefault: _formatMessage(message, values, DEFAULT_LOCALE),
+    };
   };
 
-  return getMessageInstanceIdFn;
+  return getIcuMessageFn;
 }
 
 /**
- * Returns true if string is an ICUMessage reference.
- * @param {string} icuMessageIdOrRawString
- * @return {boolean}
+ * Returns whether `icuMessageOrNot`` is an `LH.IcuMessage` instance.
+ * @param {unknown} icuMessageOrNot
+ * @return {icuMessageOrNot is LH.IcuMessage}
  */
-function isIcuMessage(icuMessageIdOrRawString) {
-  return MESSAGE_INSTANCE_ID_QUICK_REGEX.test(icuMessageIdOrRawString) &&
-      MESSAGE_INSTANCE_ID_REGEX.test(icuMessageIdOrRawString);
-}
+function isIcuMessage(icuMessageOrNot) {
+  if (!isObjectOfUnknownValues(icuMessageOrNot)) {
+    return false;
+  }
 
-/**
- * @param {string} icuMessageIdOrRawString
- * @param {LH.Locale} locale
- * @return {string}
- */
-function getFormatted(icuMessageIdOrRawString, locale) {
-  if (isIcuMessage(icuMessageIdOrRawString)) {
-    return _resolveIcuMessageInstanceId(icuMessageIdOrRawString, locale).formattedString;
+  const {i18nId, values, formattedDefault} = icuMessageOrNot;
+  if (typeof i18nId !== 'string') {
+    return false;
   }
 
-  return icuMessageIdOrRawString;
-}
+  // formattedDefault is required.
+  if (typeof formattedDefault !== 'string') {
+    return false;
```

Comment:
way to go codecov, it's like a burndown list of test cases when it's a utility module 👍